crates/etl/src/lib.rs

```diff
@@ -0,0 +1,256 @@
+    buffer_capacity_bytes: usize,
+    /// In-memory buffer storing encoded and compressed key-value pairs
+    buffer: Vec<(<K as Encode>::Encoded, <V as Compress>::Compressed)>,
+    /// Total number of elements in the collector, including all files
+    len: usize,
+}
+
+impl<K, V> Collector<K, V>
+where
+    K: Key,
+    V: Value,
+    <K as Encode>::Encoded: Ord + std::fmt::Debug,
+    <V as Compress>::Compressed: Ord + std::fmt::Debug,
+{
+    /// Create a new collector in a specific temporary directory with some capacity.
+    ///
+    /// Once the capacity (in bytes) is reached, the data is sorted and flushed to disk.
+    pub fn new(dir: Arc<TempDir>, buffer_capacity_bytes: usize) -> Self {
+        Self {
+            dir,
+            buffer_size_bytes: 0,
+            files: Vec::new(),
+            buffer_capacity_bytes,
+            buffer: Vec::new(),
+            len: 0,
+        }
+    }
+
+    /// Returns number of elements currently in the collector.
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns `true` if there are currently no elements in the collector.
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Insert an entry into the collector.
+    pub fn insert(&mut self, key: K, value: V) {
+        let key = key.encode();
+        let value = value.compress();
+        self.buffer_size_bytes += key.as_ref().len() + value.as_ref().len();
+        self.buffer.push((key, value));
+        if self.buffer_size_bytes > self.buffer_capacity_bytes {
+            self.flush();
+        }
+        self.len += 1;
+    }
+
+    fn flush(&mut self) {
+        self.buffer_size_bytes = 0;
+        self.buffer.sort_unstable_by(|a, b| a.0.cmp(&b.0));
+        let mut buf = Vec::with_capacity(self.buffer.len());
+        std::mem::swap(&mut buf, &mut self.buffer);
+        self.files.push(EtlFile::new(self.dir.path(), buf).expect("could not flush data to disk"))
+    }
+
+    /// Returns an iterator over the collector data.
+    ///
+    /// The items of the iterator are sorted across all underlying files.
+    ///
+    /// # Note
+    ///
+    /// The keys and values have been pre-encoded, meaning they *SHOULD NOT* be encoded or
+    /// compressed again.
+    pub fn iter(&mut self) -> std::io::Result<EtlIter<'_>> {
+        // Flush the remaining items to disk
+        if self.buffer_size_bytes > 0 {
+            self.flush();
+        }
+
+        let mut heap = BinaryHeap::new();
+        for (current_id, file) in self.files.iter_mut().enumerate() {
+            if let Some((current_key, current_value)) = file.read_next()? {
+                heap.push((Reverse((current_key, current_value)), current_id));
+            }
+        }
+
+        Ok(EtlIter { heap, files: &mut self.files })
+    }
+}
+
+/// An iterator over sorted data in a collection of ETL files.
+#[derive(Debug)]
+pub struct EtlIter<'a> {
+    /// Heap managing the next items to be iterated.
```

Comment:
I'd like more context here what the vec's data is and what exactly the this iterator yields